### PR TITLE
Increase the default window size to 1024×576, maximum to 1440x840

### PR DESCRIPTION
**Pull request by felipeerias**
Use 1024×576 as the default size for Web browser windows.

All widely used CSS frameworks have a breakpoint around 1024px
so they can target tablets and regular desktop windows.

The aspect ratio of the default window remains 16:9. 

The maximum size is 1440×810, which preserves the same aspect ratio
while remaining (more or less) readable.

Between the default and maximum values, we will scale the windows
proportionally.

### CODE REVIEW

**Comment by felipeerias** [comment]
These are the sizes for each scale value:

- scale **0.5** : **707×424** (50% of the area of the default window, the same calculation as before)
- scale **1.0** : **1000×600** (default)
- scale **1.5** : **1100×660**
- scale **2.0** : **1200×720**
- scale **3.0** : **1400×840** (maximum)

You can use this page for testing: https://people.igalia.com/femorandeira/viewport/

File: app/src/main/cpp/Cylinder.cpp [hunk 1]
```diff
@@ -28,6 +28,8 @@ namespace crow {
 // Ratio between world size and cylinder surface size.
 // It should match the values defined in WindowWidget.
 // 800px is the default window size for a 4m world size.
+// TODO: For now, this value remains the same even if the default window size changes
+//       because otherwise the default windows are not correctly mapped on the cylinder.
```
Comment (svillar) [inline app/src/main/cpp/Cylinder.cpp:32]: what do you mean here with the "Default windows" ?

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/WindowWidget.java [hunk 2]
```diff
@@ -1493,11 +1496,27 @@ public float getMaxWindowScale() {
     }
 
     public @NonNull Pair<Float, Float> getSizeForScale(float aScale, float aAspect) {
-        float worldWidth = WidgetPlacement.floatDimension(getContext(), R.dimen.window_world_width) *
-                (float)SettingsStore.getInstance(getContext()).getWindowWidth() / (float)SettingsStore.WINDOW_WIDTH_DEFAULT;
-        float worldHeight = worldWidth / aAspect;
-        float area = worldWidth * worldHeight * aScale;
-        float targetWidth = (float) Math.sqrt(area * aAspect);
+        final float defaultWidthWorld = WidgetPlacement.floatDimension(getContext(), R.dimen.window_world_width);
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/WindowWidget.java:1499]: this should be `defaultWorldWidth`

**Review by svillar — CHANGES_REQUESTED** [review]
Looks like a step in the right direction, I'll still need to test it on a variety of devices.

As a general comment we need better commit messages. They describe what the patch does but we need an explanation of the issue and a clear justification

**Comment by javifernandez** [comment]
> These are the sizes for each scale value:
> 
>     * scale **0.5** : **707×424** (50% of the area of the default window, the same calculation as before)
> 
>     * scale **1.0** : **1000×600** (default)
> 
>     * scale **1.5** : **1100×660**
> 
>     * scale **2.0** : **1200×720**
> 
>     * scale **3.0** : **1400×840** (maximum)
> 
> 
> You can use this page for testing: https://people.igalia.com/femorandeira/viewport/

How this change affects to the test case defined in the [issue#604](https://github.com/Igalia/wolvic/issues/604) ?

**Comment by svillar** [comment]
Apart from the specific review comments I do have a question: what specific use cases does this exactly fix/improve?

**Comment by javifernandez** [comment]
> Apart from the specific review comments I do have a question: what specific use cases does this exactly fix/improve?

I hope it could fix the [issue #604](https://github.com/Igalia/wolvic/issues/604); if not, I'd suggest to analyze how it related to this change and what would be needed to properly fix it.

File: app/src/main/cpp/Cylinder.cpp (hunk 1 above)
Comment (felipeerias) [reply app/src/main/cpp/Cylinder.cpp:32]: This value remains the same even if the default window size changes to ensure that three windows at the default size cover 180° of the cylinder.

If I increase this value following the increase in default window size, those three windows cover more than half of the cylinder.

Comment (felipeerias) [reply app/src/main/cpp/Cylinder.cpp:32]: Comment updated:

```java
// TODO: For now, this value remains the same even if the default window size changes
//       to ensure that three windows at the default size cover 180° of the cylinder
//       (which is not the case if we increase this value).
```

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/WindowWidget.java (hunk 2 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/WindowWidget.java:1499]: Done.

**Comment by felipeerias** [comment]
@javifernandez  This is not really related to https://github.com/Igalia/wolvic/issues/604, which keeps happening with and without this patch 😕

Dailymotion changes the layout at 1024 px. That seems a popular value for a breakpoint and several CSS frameworks use it: [Bulma](https://bulma.io/documentation/overview/responsiveness/), [Foundation](https://get.foundation/sites/docs/media-queries.html#default-media-queries), [Open Props](https://open-props.style/#media-queries), [Tailwind](https://tailwindcss.com/docs/screens).

I will experiment with setting the default window size at 1024 pixels wide. Perhaps that's a better default.

**Comment by svillar** [comment]
What's the relationship of this with `https://github.com/Igalia/wolvic/issues/498#issuecomment-1463204483` ? Instead of actually increasing the size of the window perhaps we just want to start with a different viewport size

**Comment by svillar** [comment]
> What's the relationship of this with `https://github.com/Igalia/wolvic/issues/498#issuecomment-1463204483` ? Instead of actually increasing the size of the window perhaps we just want to start with a different viewport size

I'd like to know the answer to this before merging. Also you mentioned that 1024 is probably a better size, shouldn't we use it?

**Comment by felipeerias** [comment]
### DPI

DPI usually defines how many physicals device pixels are in one CSS pixel. However, in our case it is used to maps CSS pixels to pixels in a texture.

From my point of view, we should keep the DPI at 1.0 by default. Users can modify this value to make websites look bigger or smaller, although the resulting texture might look blurry : e.g. with `dpi=0.75` 3 CSS pixels would be squeezed into 2 texture pixels.

**Comment by felipeerias** [comment]
### Window size

In Wolvic, _Window_ width and height in pixels defines the size of the texture where the website will be rendered.

_World_ width and height define the size of the website texture in the 3D VR world.

Currently, the default _window_ width of 800px is equivalent to a _world_ width of 4.0.

My proposal is to set the default _window_ size to **1024x576px** while keeping the same _world_ width of 4.0.

In practice, this means that websites are zoomed out by 28% while the proportion of the default window remains **16:9**.

Later on, we can explore whether this mechanism could be a viable way to implement zooming without relying on the Web engine.

The value of 1024 px comes from reviewing the breakpoint sizes in popular CSS frameworks:

| Bulma | Foundation | OpenProps | Tailwind | Bootstrap | MaterialUI | SemanticUI | UIkit | Materialize | PureCSS |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
|  769 |  640 |  768 |  768 |  768 |  600 |  723 |  768 |  600 |  768 |
| 1024 | 1024 | 1024 | 1024 |  992 |  900 |  933 |  960 |  992 | 1024 |
| 1216 | 1200 |	1440 | 1280 | 1200 | 1200 | 1127 | 1200 | 1200 | 1280 |
| 1408 | 1440 | 1920 | 1536 |      |      |      |      |      | 1920 | 

As you can see, all of them have a breakpoint around 1024 px to target tables and regular desktop windows. Experimenting with different websites also show that they are more readable at 1024px.

I am also setting the maximum window size to **1440x840px**, which preserves the same aspect ratio
while remaining (more or less) readable.

Our resizing mechanism basically tried to increase and decrease the total area of the window, which did not always result in useful sizes. I have modified this slightly so it will simply scale windows proportionally between the default and maximum points. This means that the steps become:

- _0.5x_ : 724 × 407
- _1.0x_ : 1024 × 576
- _1.5x_ : 1128 × 634
- _2.0x_ : 1232 × 693
- _3.0x_ : 1440 × 810

Finally, if this change goes through we might reconsider whether we really need desktop mode, which uses a smaller fixed width of 980 CSS px.

File: app/src/common/shared/com/igalia/wolvic/browser/SettingsStore.java [hunk 3]
```diff
@@ -85,11 +85,11 @@ SettingsStore getInstance(final @NonNull Context aContext) {
     public final static int UA_MODE_DEFAULT = WSessionSettings.USER_AGENT_MODE_VR;
     public final static int INPUT_MODE_DEFAULT = 1;
     public final static float DISPLAY_DENSITY_DEFAULT = 1.0f;
-    public final static int WINDOW_WIDTH_DEFAULT = 800;
-    public final static int WINDOW_HEIGHT_DEFAULT = 450;
+    public final static int WINDOW_WIDTH_DEFAULT = 1024;
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/browser/SettingsStore.java:88]: I wonder whether we should rename these to WINDOW_TEX_[WIDTH|HEIGHT]_DEFAULT so we don't have to review everytime which width|height are we talking about.

File: app/src/main/cpp/Cylinder.cpp [hunk 4]
```diff
@@ -28,6 +28,9 @@ namespace crow {
 // Ratio between world size and cylinder surface size.
 // It should match the values defined in WindowWidget.
 // 800px is the default window size for a 4m world size.
+// TODO: For now, this value remains the same even if the default window size changes
+//       to ensure that three windows at the default size cover 180° of the cylinder
+//       (which is not the case if we increase this value).
```
Comment (svillar) [inline app/src/main/cpp/Cylinder.cpp:33]: The fact that we don't have to change this makes me thing that there is something that we don't really control. Perhaps the native code is not respecting this value somewhere but it's weird that it does not change. Mind double checking again why?

**Comment by felipeerias** [comment]
Replaced by https://github.com/Igalia/wolvic/pull/1105

**Pull request closed without merging**